tket/src/WeightSubgrMono/EndToEndWrappers/MainSolver.cpp

```diff
@@ -0,0 +1,421 @@
+      return;
+    }
+
+    // Now we can search!
+    ++m_solution_data.iterations;
+
+    // On the first move ONLY, we don't backtrack; but we also haven't reduced.
+    if (m_solution_data.iterations == 1) {
+      if (!m_search_branch_ptr->reduce_current_node(reduction_parameters)) {
+        m_solution_data.finished = true;
+        return;
+      }
+    } else {
+      if (!m_search_branch_ptr->backtrack(reduction_parameters)) {
+        m_solution_data.finished = true;
+        return;
+      }
+    }
+    if (move_down_from_reduced_node(reduction_parameters)) {
+      // We've GOT a complete solution! Note that it MUST be good enough
+      // to add, since we've set the max weight already.
+      // We also already checked that we haven't yet got too many,
+      // if we're storing more than one.
+      add_solution_from_final_node(parameters, reduction_parameters);
+      if (terminate_with_enough_full_solutions(parameters, m_solution_data)) {
+        return;
+      }
+    }
+    if (Clock::now() >= desired_end_time) {
+      return;
+    }
+  }
+}
+
+void MainSolver::add_solution_from_final_node(
+    const MainSolverParameters& parameters,
+    const SearchBranch::ReductionParameters& reduction_parameters) {
+  TKET_ASSERT(m_pre_search_components_ptr);
+  TKET_ASSERT(m_search_branch_ptr);
+
+  const auto& accessor = m_search_branch_ptr->get_domains_accessor();
+  const auto scalar_product = accessor.get_scalar_product();
+
+  TKET_ASSERT(
+      accessor.get_total_p_edge_weights() ==
+      m_solution_data.total_p_edge_weights);
+
+  TKET_ASSERT(scalar_product <= reduction_parameters.max_weight);
+
+  // We'll overwrite the solution into back().
+  if (parameters.for_multiple_full_solutions_the_max_number_to_obtain > 0 ||
+      m_solution_data.solutions.empty()) {
+    m_solution_data.solutions.emplace_back();
+  }
+  auto& assignments = m_solution_data.solutions.back().assignments;
+  const auto number_of_pv = accessor.get_number_of_pattern_vertices();
+  assignments.clear();
+  assignments.reserve(number_of_pv);
+
+  for (unsigned pv = 0; pv < number_of_pv; ++pv) {
+    const auto& domain = accessor.get_domain(pv);
+    TKET_ASSERT(domain.size() == 1);
+    assignments.emplace_back(pv, *domain.cbegin());
+  }
+  m_solution_data.solutions.back().scalar_product = scalar_product;
+
+  m_solution_data.solutions.back().total_p_edges_weight =
+      m_solution_data.total_p_edge_weights;
+}
+
+bool MainSolver::move_down_from_reduced_node(
+    const SearchBranch::ReductionParameters& reduction_parameters) {
+  TKET_ASSERT(m_search_components_ptr);
+  TKET_ASSERT(m_search_branch_ptr);
+
+  for (;;) {
+    const auto next_var_result =
+        m_search_components_ptr->variable_ordering.get_variable(
+            m_search_branch_ptr->get_domains_accessor_nonconst(),
+            m_search_components_ptr->rng);
+
+    if (next_var_result.empty_domain) {
+      return false;
+    }
+    if (!next_var_result.variable_opt) {
```

Comment:
Comment added; in next push!

"...if there are no empty domains and no variable assignments then it _is _ a full assignment?"

Yes, exactly. An empty domain means a nogood (the partial solution cannot be extended to a full solution).

`next_var_result` comes from `variable_ordering.get_variable`. Thus the `variable_ordering` is responsible for choosing which PV should be assigned next.

(Then, a different class `ValueOrdering` is responsible for choosing one of the TV inside Domain(PV). Note that "variable ordering" and "value ordering" are both standard terminology in this subject).

If `variable_ordering` cannot choose a PV, it is because EVERY PV has been assigned, so it is indeed a full solution.

(Before it reached this stage, it was already checked that any new pattern edges joining each new PV to existing assigned PVs ARE indeed mapped to valid target edges, so we don't need any further validity check here).